v2/api/dataprotection/customizations/backup_vaults_backup_instance_extensions.go

```diff
@@ -0,0 +1,142 @@
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+
+	armdataprotection "github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/dataprotection/armdataprotection/v3"
+	dataprotection "github.com/Azure/azure-service-operator/v2/api/dataprotection/v1api20231101/storage"
+
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/extensions"
+)
+
+var (
+	_ extensions.PreReconciliationChecker  = &BackupVaultsBackupInstanceExtension{}
+	_ extensions.PostReconciliationChecker = &BackupVaultsBackupInstanceExtension{}
+)
+
+var nonRetryableStates = set.Make(
+	"ConfiguringProtectionFailed",
+	"Invalid",
+	"NotProtected",
+	"ProtectionConfigured",
+	"SoftDeleted",
+	"ProtectionStopped",
+	"BackupSchedulesSuspended",
+	"RetentionSchedulesSuspended",
+)
+
+var retryableStates = set.Make(
+	"ProtectionError",
+	"UpdatingProtection",
+	"ConfiguringProtection",
+	"SoftDeleting",
+)
+
+func (extension *BackupVaultsBackupInstanceExtension) PostReconcileCheck(
+	ctx context.Context,
+	obj genruntime.MetaObject,
+	owner genruntime.MetaObject,
+	_ *resolver.Resolver,
+	armClient *genericarmclient.GenericClient,
+	log logr.Logger,
+	_ extensions.PostReconcileCheckFunc,
+) (extensions.PostReconcileCheckResult, error) {
+	log.V(Debug).Info("########################## Starting Post-reconcilation for Backup Instance ##########################")
+	backupInstance, ok := obj.(*dataprotection.BackupVaultsBackupInstance)
+	if !ok {
+		return extensions.PostReconcileCheckResult{},
+			errors.Errorf("cannot run on unknown resource type %T, expected *dataprotection.BackupVaultsBackupInstance", obj)
+	}
+
+	var _ conversion.Hub = backupInstance
+
+	if owner != nil && backupInstance != nil && backupInstance.Status.Id != nil &&
+		backupInstance.Status.Properties != nil && backupInstance.Status.Properties.ProtectionStatus != nil &&
+		backupInstance.Status.Properties.ProtectionStatus.Status != nil {
+
+		protectionStatus := *backupInstance.Status.Properties.ProtectionStatus.Status
+		log.V(Debug).Info(fmt.Sprintf("########################## Protection Status is  %q ##########################", protectionStatus))
+
+		// return success for reconcilation if the state is non-retryable
+		if nonRetryableStates.Contains(protectionStatus) {
+			log.V(Debug).Info("########################## Returning PostReconcileCheckResultSuccess ##########################")
+			return extensions.PostReconcileCheckResultSuccess(), nil
+		}
+
+		// call sync api only when protection status is ProtectionError and error code is usererror
+		if protectionStatus == "ProtectionError" {
+			protectionStatusErrorCode := ""
+			protectionStatusErrorCode = strings.ToLower(*backupInstance.Status.Properties.ProtectionStatus.ErrorDetails.Code)
```

Comment:
No need to declare and then immediately overwrite, just initialize it directly.
```suggestion
			protectionStatusErrorCode := strings.ToLower(*backupInstance.Status.Properties.ProtectionStatus.ErrorDetails.Code)
```

Reply:
actually I needed to move it above if condition